### PR TITLE
[core] Add RuleViolationFix interface to generate autofixes && Add CLI flag to ignore cache

### PR DESCRIPTION
**Pull request by gibarsin**
# Summary

Implements #693:
- Add an interface to implement autofixes per rule
- Add autofix command line parameter to run rules with autofixes (currently ignored)

1) There needs to be a way to point to the object which will attempt to fix a reported rule violation. When a rule violation is detected while traversing the AST, an optional auto fixer class, along with the node on which it is desired to operate on, should be reported along with the violation.
2) Currently incremental analysis is not supported because auto fixes are not cached. Therefore when attempting to apply auto fixes over an AST + Document the cache must be ignored. The flag is currently ignored.

# Additions
- Add **RuleViolationFix** interface to operate on a node either to add another node, or to modify/delete the current node.
- Add **AutoFixableRuleViolation** to retrieve the fixer class for a detected rule violation. Note that because a Rule can report more than one type of rule violation, no association whatsoever is enforce between a Rule and the AutoFixableRuleViolation node type.
- Add methods to **RuleViolationFactory** and **AbstractRuleViolationFactory** to deal with the report of rule violations with intention of auto fix them.
- Add CLI flag to diff when attempting to apply autofixes. This allows for ignoring the cache until it is supported in a future PR.

### CODE REVIEW

File: pmd-core/src/main/java/net/sourceforge/pmd/lang/rule/AbstractRuleViolationFactory.java [hunk 1]
```diff
@@ -39,16 +41,42 @@ public void addViolation(RuleContext ruleContext, Rule rule, Node node, String m
...
+     * @param ruleContext the context of the rule at the moment of the violation
+     * @param node the node in which the fixes to that rule violation will be applied
+     * @param message the message to put in the report
+     * @param autoFixerClass the class in which the fixes to that rule violation will be applied
+     * @return a RuleViolation with the capability of telling its fixer class
+     */
+    protected AutoFixableRuleViolation createRuleViolationWithAutoFixer(final Rule rule, final RuleContext ruleContext,
+                                                                        final Node node, final String message,
+                                                                        final Class<? extends RuleViolationFix> autoFixerClass) {
+        throw new UnsupportedOperationException("Subclasses of AbstractRuleViolationFactory should override "
```
[the remark this answers is not in the document]
Comment (gibarsin) [reply pmd-core/src/main/java/net/sourceforge/pmd/lang/rule/AbstractRuleViolationFactory.java]: The reason is commented in the Javadocs

File: pmd-core/src/main/java/net/sourceforge/pmd/lang/rule/RuleViolationFactory.java [hunk 2]
```diff
@@ -31,4 +32,16 @@
...
+    /**
+     * Create a rule violation with the appropriate fixer. Note that the rule is dissociated from the fixer class.
+     *
+     * @param ruleContext the context of the rule at the moment of the violation
+     * @param rule the rule which reported the violation
+     * @param node the node in which the fixes to that rule violation will be applied
+     * @param message the message to put in the report
+     * @param autoFixerClass the class in which the fixes to that rule violation will be applied
+     */
+    void addViolationWithAutoFixer(RuleContext ruleContext, Rule rule, Node node, String message,
```
[the remark this answers is not in the document]
Comment (gibarsin) [reply pmd-core/src/main/java/net/sourceforge/pmd/lang/rule/RuleViolationFactory.java]: I'll add it

File: pmd-core/src/main/java/net/sourceforge/pmd/lang/rule/AbstractRuleViolationFactory.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (jsotuyod) [reply pmd-core/src/main/java/net/sourceforge/pmd/lang/rule/AbstractRuleViolationFactory.java]: That is not a valid reason. All the work on AST editing and document editing is at pmd-core level, which means it's free for all.

File: pmd-core/src/main/java/net/sourceforge/pmd/SourceCodeProcessor.java [hunk 3]
```diff
@@ -82,9 +78,8 @@ public void processSourceCode(Reader sourceCode, RuleSets ruleSets, RuleContext
 
         // Coarse check to see if any RuleSet applies to file, will need to do a finer RuleSet specific check later
         if (ruleSets.applies(ctx.getSourceCodeFile())) {
-            // Is the cache up to date?
-            if (configuration.getAnalysisCache().isUpToDate(ctx.getSourceCodeFile())) {
-                for (final RuleViolation rv : configuration.getAnalysisCache().getCachedViolations(ctx.getSourceCodeFile())) {
+            if (!configuration.isAutoFixes() && isCacheUpToDate(ctx)) {
```
[the remark this answers is not in the document]
Comment (gibarsin) [reply pmd-core/src/main/java/net/sourceforge/pmd/SourceCodeProcessor.java]: I know what you mean, but currently we do not even store the class for the fixer in any rule violation, so we cannot know if there were fixes to apply, so if the user wants autofixes, he needs to recheck the files. These issue will be addressed in the following days when cache for autofixes is added, avoiding this conditional all together.

File: pmd-core/src/main/java/net/sourceforge/pmd/SourceCodeProcessor.java [hunk 4]
```diff
@@ -105,6 +100,14 @@ public void processSourceCode(Reader sourceCode, RuleSets ruleSets, RuleContext
         }
     }
 
+    private boolean isCacheUpToDate(final RuleContext context) {
+        return configuration.getAnalysisCache().isUpToDate(context.getSourceCodeFile());
+    }
+
+    private Iterable<RuleViolation> getCachedViolationsForFile(final File sourceCodeFile) {
```
[the remark this answers is not in the document]
Comment (gibarsin) [reply pmd-core/src/main/java/net/sourceforge/pmd/SourceCodeProcessor.java]: I added it for enhanced readability, but will revert.

File: pmd-core/src/main/java/net/sourceforge/pmd/cli/PMDParameters.java [hunk 5]
```diff
@@ -102,6 +102,9 @@
     @Parameter(names = "-cache", description = "Specify the location of the cache file for incremental analysis.")
     private String cacheLocation = null;
 
+    @Parameter(names = "-fix", description = "Attempt to automatically fix rule violations found on source files.")
+    private boolean autoFixes = false;
```
[the remark this answers is not in the document]
Comment (gibarsin) [reply pmd-core/src/main/java/net/sourceforge/pmd/cli/PMDParameters.java]: Ok, updating the PR description too.

File: pmd-core/src/main/java/net/sourceforge/pmd/lang/rule/AbstractRule.java [hunk 6]
```diff
@@ -395,6 +395,12 @@ public void addViolationWithMessage(Object data, Node node, String message, Obje
                 this, node, message, args);
     }
 
+    public void addViolationWithAutoFixerClass(Object data, Node node, String message, Class autoFixerClass) {
```
[the remark this answers is not in the document]
Comment (gibarsin) [reply pmd-core/src/main/java/net/sourceforge/pmd/lang/rule/AbstractRule.java]: It slipped my mind.

[the remark this answers is not in the document]
Comment (gibarsin) [reply pmd-core/src/main/java/net/sourceforge/pmd/lang/rule/AbstractRule.java]: I'll change it

File: pmd-core/src/main/java/net/sourceforge/pmd/lang/rule/AbstractRuleViolationFactory.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (gibarsin) [reply pmd-core/src/main/java/net/sourceforge/pmd/lang/rule/AbstractRuleViolationFactory.java]: By free for all you mean that implementations should be given at implementation level of the factories? If that's the case I should modify each language's implementation of RuleViolation to support auto fixes, or to copy this same implementation (throwing an exception) if it's not currently supported.

File: pmd-core/src/main/java/net/sourceforge/pmd/lang/rule/AbstractRuleViolationFactory.java [hunk 7]
```diff
@@ -39,16 +41,42 @@ public void addViolation(RuleContext ruleContext, Rule rule, Node node, String m
...
 
         String formattedMessage = cleanup(message, args);
 
         ruleContext.getReport()
                 .addRuleViolation(createRuleViolation(rule, ruleContext, node, formattedMessage, beginLine, endLine));
     }
 
-    protected abstract RuleViolation createRuleViolation(Rule rule, RuleContext ruleContext, Node node, String message);
+    @Override
+    public void addViolationWithAutoFixer(final RuleContext ruleContext, final Rule rule, final Node node,
```
[the remark this answers is not in the document]
Comment (gibarsin) [reply pmd-core/src/main/java/net/sourceforge/pmd/lang/rule/AbstractRuleViolationFactory.java]: I'll change them

File: pmd-core/src/main/java/net/sourceforge/pmd/lang/rule/AbstractRuleViolationFactory.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (jsotuyod) [reply pmd-core/src/main/java/net/sourceforge/pmd/lang/rule/AbstractRuleViolationFactory.java]: Implementing the method is a one-liner to instantiate the proper type / parametric rule violation, it's 0-cost for languages to support it. There is no reason to default to this. Moreover, doing it this way would not fail upon execution with a clear message, but would process all files and simply produce a bunch of `ProcessingError`s for `UnsupportedOperationException`

**Comment by gibarsin** [comment]
Regarding testing, I'm hesitant about whether there can be useful tests here, since the only functionality this PR provides is the assignment of rule violation fix classes to classes that extend from existing RuleViolation classes.

File: pmd-core/src/main/java/net/sourceforge/pmd/autofix/RuleViolationFix.java [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+/**
+ * Classes of this interface must be used when it is desired to make one or more operations to the AST.
+ */
+public interface RuleViolationFix {
+
+    /**
+     * Apply one or more operations to a node.
+     * @param node the node in the AST on which to apply operations
+     */
+    void applyFixesToNode(Node node);
```
Comment (jsotuyod) [inline pmd-core/src/main/java/net/sourceforge/pmd/autofix/RuleViolationFix.java]: a fix applies several fixes? maybe this should simply be `applyToNode`

File: pmd-core/src/main/java/net/sourceforge/pmd/autofix/rule/DeleteRuleViolationFix.java [hunk 9]
```diff
@@ -0,0 +1,16 @@
+/**
+ * BSD-style license; for more info see http://pmd.sourceforge.net/license.html
+ */
+
+package net.sourceforge.pmd.autofix.rule;
```
Comment (jsotuyod) [inline pmd-core/src/main/java/net/sourceforge/pmd/autofix/rule/DeleteRuleViolationFix.java]: not sure why this is in a "rule" child package from the interface...

File: pmd-core/src/main/java/net/sourceforge/pmd/cache/CachedRuleViolation.java [hunk 10]
```diff
@@ -14,7 +14,7 @@
 /**
  * A {@link RuleViolation} implementation that is immutable, and therefore cache friendly
  */
-public final class CachedRuleViolation implements RuleViolation {
+public class CachedRuleViolation implements RuleViolation {
```
Comment (jsotuyod) [inline pmd-core/src/main/java/net/sourceforge/pmd/cache/CachedRuleViolation.java]: if you are not dealing with cache now, please revert these changes. On their own they make little sense

File: pmd-jsp/src/main/java/net/sourceforge/pmd/lang/jsp/rule/JspRuleViolationFactory.java [hunk 11]
```diff
@@ -25,6 +28,16 @@ protected RuleViolation createRuleViolation(Rule rule, RuleContext ruleContext,
...
     }
 
+    @Override
+    protected AutoFixableRuleViolation createRuleViolation(Rule rule, RuleContext ruleContext, Node node, String message, Class<? extends RuleViolationFix> ruleViolationFixClass) {
+        return new AutoFixableParametricRuleViolation<>(rule, ruleContext, (JspNode) node, message, ruleViolationFixClass);
+    }
+
+    @Override
+    protected AutoFixableRuleViolation createRuleViolation(Rule rule, RuleContext ruleContext, Node node, String message, int beginLine, int endLine, Class<? extends RuleViolationFix> ruleViolationFixClass) {
+        return null;
```
Comment (jsotuyod) [inline pmd-jsp/src/main/java/net/sourceforge/pmd/lang/jsp/rule/JspRuleViolationFactory.java]: this should be implementable the same way you did for other languages such as PLSQL

File: pmd-javascript/src/main/java/net/sourceforge/pmd/lang/ecmascript/rule/EcmascriptRuleViolationFactory.java [hunk 12]
```diff
@@ -25,6 +28,16 @@ protected RuleViolation createRuleViolation(Rule rule, RuleContext ruleContext,
...
     }
 
+    @SuppressWarnings("rawtypes")
+    @Override
+    protected AutoFixableRuleViolation createRuleViolation(Rule rule, RuleContext ruleContext, Node node, String message, Class<? extends RuleViolationFix> ruleViolationFixClass) {
+        return new AutoFixableParametricRuleViolation<>(rule, ruleContext, node, message, ruleViolationFixClass);
+    }
+
+    protected AutoFixableRuleViolation createRuleViolation(Rule rule, RuleContext ruleContext, Node node, String message, int beginLine, int endLine, Class<? extends RuleViolationFix> ruleViolationFixClass) {
+        return null;
```
Comment (jsotuyod) [inline pmd-javascript/src/main/java/net/sourceforge/pmd/lang/ecmascript/rule/EcmascriptRuleViolationFactory.java]: this should be implementable the same way you did for other languages such as PLSQL

File: pmd-core/src/main/java/net/sourceforge/pmd/autofix/rule/DeleteRuleViolationFix.java (hunk 9 above)
Comment (gibarsin) [reply pmd-core/src/main/java/net/sourceforge/pmd/autofix/rule/DeleteRuleViolationFix.java]: I deleted the DeleteRuleViolationFix class for the time being. However, I'm not yet sure where rule violation fixes which apply to any language can be placed.

**Comment by gibarsin** [comment]
We are closing this as it has been deprecated in favor of a new autofixes approach. See #693.
